File: Interv01/05.py
# https://leetcode.cn/problems/one-away-lcci/?favorite=xb9lfcwi

class Solution:
    def oneEditAway(self, first: str, second: str) -> bool:
        if abs(len(first) - len(second)) > 1:
            return False
        if first == second:
            return True
        # 思路：双指针法 - 前后双指针寻找字符不匹配位置，若位置不同则不对
        if len(first) < len(second):
            first, second = second, first
        if second == "":
            return True
        start, end = 0, -1

        while start < len(second) and first[start] == second[start]:
            start += 1
        while abs(end) <= len(second) and first[end] == second[end]:
            end -= 1

        if start == len(first) + end:
            return True
        else:
            return False

if __name__ == '__main__':
    test = Solution()
    first = "b"
    second = "ab"
    print(test.oneEditAway(first, second))

File: Interv01/05.cpp
//
// Created by Air on 2023/2/26.
//
// https://leetcode.cn/problems/one-away-lcci/

#include "iostream"
using namespace std;

class Solution {
public:
    bool oneEditAway(string first, string second) {
        if (abs(int(first.size() - second.size())) > 1)
            return false;
        if (first == second)
            return true;

        if (first.size() < second.size())
            std::swap(first, second);

        int first_p(0), second_p(0);
        int first_q(first.size()-1), second_q(second.size()-1);

        while (first_p < first_q && first[first_p] == second[second_p] ) {
            ++first_p;
            ++second_p;
        }
        while (first_p < first_q && first[first_q] == second[second_q]) {
            --first_q;
            --second_q;
        }

        return first_p == first_q;

    }
};

int main() {
    Solution test;
    string s1("bcd");
    string s2("abcd");
    cout << test.oneEditAway(s1, s2) << endl;
    return 0;
}